README.md
# Certify

Create client certificates for your TLS app.

```rust
let ca = load_ca(ca_cert_pem, ca_key_pem)?;
let (cert_pem, key_pem) = generate_cert(&ca, &[], "US", "web", "abcd1234", true, 365)?;
```

## License

`certify` is distributed under the terms of MIT.

See [LICENSE](LICENSE.md) for details.

Copyright 2021 Tyr Chen


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.5.1] - 2022-11-26

[c9dc20b](c9dc20b797c8e405e3fe938166eb56036293a185)...[b91f245](b91f2458cbd9d56d17461930aa4b20386a5e31ba)

### Features

- Allow client to choose which sig algo to use (osx keychain does not support ed25519 yet) ([b91f245](b91f2458cbd9d56d17461930aa4b20386a5e31ba) - 2022-11-26 by Tyr Chen)

## [0.5.0] - 2022-11-26

[e2631f6](e2631f602e8c5c94fba46c280491c287e883ffe9)...[c9dc20b](c9dc20b797c8e405e3fe938166eb56036293a185)

### Bug Fixes

- Use time crate not chrono crate ([c9dc20b](c9dc20b797c8e405e3fe938166eb56036293a185) - 2022-11-26 by Tyr Chen)

### Miscellaneous Tasks

- Upgrade deps ([6aef7d4](6aef7d412c48c7ef5f5e1a905b3da405b0a27d05) - 2022-11-26 by Tyr Chen)

<!-- generated by git-cliff -->


src/lib.rs
mod cert;
mod error;

pub use cert::{CertInfo, CertSigAlgo, CertType, CA};
pub use error::CertifyError;

// re-exports
pub use rcgen::KeyPair;

/// Generate CA cert
pub fn generate_ca(
    country: &str,
    org: &str,
    cn: &str,
    sig_algo: CertSigAlgo,
    pem_str: Option<&str>,
    days: Option<i64>,
) -> Result<(String, String), CertifyError> {
    let params = CertInfo::new(
        Vec::<String>::new(),
        Vec::<String>::new(),
        country,
        org,
        cn,
        days,
        sig_algo,
    );
    let keypair = match pem_str {
        Some(v) => Some(KeyPair::from_pem(v)?),
        None => None,
    };
    let ca = params.ca_cert(keypair)?;
    Ok((ca.serialize_pem().unwrap(), ca.serialize_private_key_pem()))
}

/// generate cert signed by the CA
#[allow(clippy::too_many_arguments)]
pub fn generate_cert(
    ca: &CA,
    domains: Vec<impl Into<String>>,
    country: &str,
    org: &str,
    cn: &str,
    sig_algo: CertSigAlgo,
    pem_str: Option<&str>,
    is_client: bool,
    days: Option<i64>,
) -> Result<(String, String), CertifyError> {
    let params = CertInfo::new(
        domains,
        vec!["127.0.0.1", "::1"],
        country,
        org,
        cn,
        days,
        sig_algo,
    );
    let keypair = match pem_str {
        Some(v) => Some(KeyPair::from_pem(v)?),
        None => None,
    };
    let cert = if is_client {
        params.client_cert(keypair)?
    } else {
        params.server_cert(keypair)?
    };
    let (cert_pem, key_pem) = ca.sign_cert(&cert)?;
    Ok((cert_pem, key_pem))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn generate_ca_cert_should_work() -> Result<(), CertifyError> {
        let (cert, key) = gen_ca(None)?;
        let ca = CA::load(&cert, &key)?;
        let cert1 = ca.serialize_pem()?;

        assert_eq!(cert, cert1);
        Ok(())
    }

    #[test]
    fn generate_ca_cert_with_existing_key_should_work() -> Result<(), CertifyError> {
        let key_pem = include_str!("fixtures/ca_key.pem");
        let ca_pem = include_str!("fixtures/ca_cert.pem");
        let (cert, key) = gen_ca(Some(key_pem))?;
        let ca = CA::load(&cert, &key)?;
        let cert1 = ca.serialize_pem()?;

        assert_eq!(cert, cert1);
        assert_eq!(key_pem, key);

        let ca = CA::load(ca_pem, &key)?;
        let cert2 = ca.serialize_pem()?;

        assert_eq!(ca_pem, cert2);

        Ok(())
    }

    #[test]
    fn generate_server_cert_with_ca_should_work() -> Result<(), CertifyError> {
        let (cert, key) = gen_ca(None)?;
        let ca = CA::load(&cert, &key)?;
        let (server_cert, server_key) = generate_cert(
            &ca,
            vec!["app.domain.com"],
            "US",
            "Domain Domain Inc.",
            "API Server",
            CertSigAlgo::ED25519,
            None,
            false,
            Some(365),
        )?;

        println!("{}\n{}", server_cert, server_key);

        Ok(())
    }

    #[test]
    fn generate_server_cert_with_existing_ca_and_key_should_work() -> Result<(), CertifyError> {
        let key_pem = include_str!("fixtures/ca_key.pem");
        let ca_pem = include_str!("fixtures/ca_cert.pem");
        let server_key_pem = include_str!("fixtures/server_key.pem");

        let ca = CA::load(ca_pem, key_pem)?;
        let (server_cert, server_key) = generate_cert(
            &ca,
            vec!["app.domain.com"],
            "US",
            "Domain Domain Inc.",
            "API Server",
            CertSigAlgo::ED25519,
            Some(server_key_pem),
            false,
            Some(365),
        )?;

        assert_eq!(&server_key, server_key_pem);

        println!("{}\n{}", server_cert, server_key);

        Ok(())
    }
    #[test]
    fn generate_client_cert_with_ca_should_work() -> Result<(), CertifyError> {
        let (cert, key) = gen_ca(None)?;
        let ca = CA::load(&cert, &key)?;
        let (client_cert, client_key) = generate_cert(
            &ca,
            vec!["app.domain.com"],
            "CA",
            "macos",
            "awesome_device_id",
            CertSigAlgo::ED25519,
            None,
            true,
            Some(365),
        )?;

        println!("{}\n{}", client_cert, client_key);

        Ok(())
    }

    #[test]
    fn generate_client_cert_with_existing_ca_and_key_should_work() -> Result<(), CertifyError> {
        let key_pem = include_str!("fixtures/ca_key.pem");
        let ca_pem = include_str!("fixtures/ca_cert.pem");
        let client_key_pem = include_str!("fixtures/client_key.pem");

        let ca = CA::load(ca_pem, key_pem)?;

        let (client_cert, client_key) = generate_cert(
            &ca,
            vec!["app.domain.com"],
            "CA",
            "macos",
            "awesome_device_id",
            CertSigAlgo::ED25519,
            Some(client_key_pem),
            true,
            Some(365),
        )?;

        assert_eq!(&client_key, client_key_pem);
        println!("{}\n{}", client_cert, client_key);

        Ok(())
    }

    fn gen_ca(pem: Option<&str>) -> Result<(String, String), CertifyError> {
        generate_ca(
            "US",
            "Domain Domain Inc.",
            "Domain CA",
            CertSigAlgo::ED25519,
            pem,
            None,
        )
    }

    // fn write_file(name: &str, content: &str) {
    //     use std::io::Write;
    //     let mut file = std::fs::File::create(name).unwrap();
    //     file.write_all(content.as_bytes()).unwrap();
    // }
}


src/cert/ca.rs
use rcgen::{Certificate, CertificateParams, KeyPair};

use super::Cert;
use crate::CertifyError;

pub struct CA {
    cert: Certificate,
    data: Option<Vec<u8>>,
}

impl CA {
    pub fn load(ca_cert: &str, ca_key: &str) -> Result<Self, CertifyError> {
        let key = KeyPair::from_pem(ca_key)?;
        let params = CertificateParams::from_ca_cert_pem(ca_cert, key)?;
        let ca_data = pem::parse(ca_cert)?.contents;
        let mut result = Self::from_params(params)?;
        result.data = Some(ca_data);
        Ok(result)
    }

    pub fn from_params(params: CertificateParams) -> Result<Self, CertifyError> {
        Ok(CA {
            cert: Certificate::from_params(params)?,
            data: None,
        })
    }

    pub fn sign_cert(&self, cert: &Cert) -> Result<(String, String), CertifyError> {
        let cert_pem = cert.0.serialize_pem_with_signer(&self.cert)?;
        let key_pem = cert.0.serialize_private_key_pem();
        Ok((cert_pem, key_pem))
    }

    pub fn serialize_der(&self) -> Result<Vec<u8>, CertifyError> {
        match &self.data {
            Some(data) => Ok(data.to_owned()),
            None => Ok(self.cert.serialize_der()?),
        }
    }

    pub fn serialize_pem(&self) -> Result<String, CertifyError> {
        let p = pem::Pem {
            tag: "CERTIFICATE".to_string(),
            contents: self.serialize_der()?,
        };
        Ok(pem::encode(&p))
    }

    pub fn serialize_private_key_pem(&self) -> String {
        self.cert.serialize_private_key_pem()
    }
}


src/cert/cert.rs
use chrono::{DateTime, Duration, NaiveDateTime, Utc};
use rcgen::{
    BasicConstraints, Certificate, CertificateParams, CustomExtension, DistinguishedName, DnType,
    ExtendedKeyUsagePurpose::*, IsCa, KeyIdMethod, KeyPair, SanType, PKCS_ED25519,
};
use std::net::IpAddr;

use super::CA;
use crate::CertifyError;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
enum Usage {
    None,
    Ca,
    Cert,
}

// const OID_ORG_UNIT: &[u64] = &[2, 5, 4, 11];
const OID_BASIC: &[u64] = &[2, 5, 29, 19];

const OID_KEY_USAGE: &[u64] = &[2, 5, 29, 15];

const KEY_USAGE: &[Usage] = &[
    Usage::Cert, // digitalSignature
    Usage::Cert, // nonRepudiation/contentCommitment
    Usage::Cert, // keyEncipherment
    Usage::None,
    Usage::None,
    Usage::Ca, // keyCertSign
    Usage::Ca, // cRLSign
    Usage::None,
    Usage::None,
];

pub struct Cert(pub Certificate);

const CERT_DEFAULT_DURATION: i64 = 180; // 180 days
const CA_DEFAULT_DURATION: i64 = 10 * 365; // approx 10 years

impl Cert {
    pub fn from_params(params: CertificateParams) -> Result<Self, CertifyError> {
        Ok(Cert(Certificate::from_params(params)?))
    }
}

#[derive(Debug, Clone)]
pub struct CertInfo {
    pub domain_names: Vec<String>,
    pub ip_address: Vec<IpAddr>,
    pub country: String,
    pub organization: String,
    pub common: String,
    pub days: Option<i64>,
}

impl CertInfo {
    pub fn new<'a, 'b>(
        domains: impl AsRef<[&'a str]>,
        ips: impl AsRef<[&'b str]>,
        country: &str,
        org: &str,
        cn: &str,
        days: Option<i64>,
    ) -> Self {
        Self {
            domain_names: domains.as_ref().iter().map(|d| d.to_string()).collect(),
            ip_address: ips.as_ref().iter().map(|ip| ip.parse().unwrap()).collect(),
            country: country.to_owned(),
            organization: org.to_owned(),
            common: cn.to_owned(),
            days,
        }
    }
}

#[derive(PartialEq)]
pub enum CertType {
    Client,
    Server,
    CA,
}

impl CertInfo {
    fn build_cert_params(
        &self,
        keypair: Option<KeyPair>,
        cert_type: CertType,
    ) -> CertificateParams {
        let alg = &PKCS_ED25519;
        let now = Utc::now().timestamp();

        let default_days = if cert_type == CertType::CA {
            CA_DEFAULT_DURATION
        } else {
            CERT_DEFAULT_DURATION
        };
        let days = self.days.unwrap_or(default_days);

        let not_before = DateTime::<Utc>::from_utc(NaiveDateTime::from_timestamp(now, 0), Utc);

        let not_after = not_before + Duration::days(days);

        let mut subject_alt_names = vec![];
        for dns in self.domain_names.iter() {
            subject_alt_names.push(SanType::DnsName(dns.to_owned()));
        }
        for ip in self.ip_address.iter() {
            subject_alt_names.push(SanType::IpAddress(ip.to_owned()));
        }
        let mut distinguished_name = DistinguishedName::new();
        distinguished_name.push(DnType::CountryName, &self.country);
        distinguished_name.push(DnType::OrganizationName, &self.organization);
        distinguished_name.push(DnType::CommonName, &self.common);

        let mut params = CertificateParams::default();
        params.key_pair = keypair;
        params.alg = alg;
        params.not_before = not_before;
        params.not_after = not_after;
        params.subject_alt_names = subject_alt_names;
        params.distinguished_name = distinguished_name;

        // WTF: turned out we should not use this to satisfy openssl
        // params.use_authority_key_identifier_extension = true;
        params.key_identifier_method = KeyIdMethod::Sha512;
        match cert_type {
            CertType::CA => {
                params.extended_key_usages = vec![Any];
                params.is_ca = IsCa::Ca(BasicConstraints::Constrained(16));
                params.custom_extensions.push(CertInfo::key_usage(true));
            }
            CertType::Client => {
                params.extended_key_usages = vec![ClientAuth];
                params.custom_extensions.push(CertInfo::not_ca());
                params.custom_extensions.push(CertInfo::key_usage(false));
            }
            CertType::Server => {
                params.extended_key_usages = vec![ServerAuth];
                params.custom_extensions.push(CertInfo::not_ca());
                params.custom_extensions.push(CertInfo::key_usage(false));
            }
        }

        params
    }

    pub fn ca_cert(&self, keypair: Option<KeyPair>) -> Result<CA, CertifyError> {
        let mut params = self.build_cert_params(keypair, CertType::CA);
        params.extended_key_usages = vec![];
        CA::from_params(params)
    }

    pub fn client_cert(&self, keypair: Option<KeyPair>) -> Result<Cert, CertifyError> {
        let params = self.build_cert_params(keypair, CertType::Client);
        Cert::from_params(params)
    }

    pub fn server_cert(&self, keypair: Option<KeyPair>) -> Result<Cert, CertifyError> {
        let params = self.build_cert_params(keypair, CertType::Server);
        Cert::from_params(params)
    }

    fn key_usage(ca: bool) -> CustomExtension {
        let der = yasna::construct_der(|writer| {
            writer.write_bitvec(
                &KEY_USAGE
                    .iter()
                    .map(|u| *u == if ca { Usage::Ca } else { Usage::Cert })
                    .collect(),
            );
        });

        let mut key_usage = CustomExtension::from_oid_content(OID_KEY_USAGE, der);
        key_usage.set_criticality(true);
        key_usage
    }

    fn not_ca() -> CustomExtension {
        let der = yasna::construct_der(|writer| {
            writer.write_sequence(|writer| {
                writer.next().write_bool(false);
            });
        });

        CustomExtension::from_oid_content(OID_BASIC, der)
    }
}


src/cert/builder.rs
use rcgen::{
    BasicConstraints, Certificate, CertificateParams, CustomExtension, DistinguishedName, DnType,
    ExtendedKeyUsagePurpose::*, IsCa, KeyIdMethod, KeyPair, SanType, SignatureAlgorithm,
    PKCS_ECDSA_P384_SHA384, PKCS_ED25519,
};
use std::net::IpAddr;
use time::{Duration, OffsetDateTime};

use super::CA;
use crate::CertifyError;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
enum Usage {
    None,
    Ca,
    Cert,
}

// const OID_ORG_UNIT: &[u64] = &[2, 5, 4, 11];
const OID_BASIC: &[u64] = &[2, 5, 29, 19];

const OID_KEY_USAGE: &[u64] = &[2, 5, 29, 15];

const KEY_USAGE: &[Usage] = &[
    Usage::Cert, // digitalSignature
    Usage::Cert, // nonRepudiation/contentCommitment
    Usage::Cert, // keyEncipherment
    Usage::None,
    Usage::None,
    Usage::Ca, // keyCertSign
    Usage::Ca, // cRLSign
    Usage::None,
    Usage::None,
];

pub struct Cert(pub Certificate);

const CERT_DEFAULT_DURATION: i64 = 180; // 180 days
const CA_DEFAULT_DURATION: i64 = 10 * 365; // approx 10 years

impl Cert {
    pub fn from_params(params: CertificateParams) -> Result<Self, CertifyError> {
        Ok(Cert(Certificate::from_params(params)?))
    }
}

#[derive(Debug, Clone)]
pub struct CertInfo {
    pub domain_names: Vec<String>,
    pub ip_address: Vec<IpAddr>,
    pub country: String,
    pub organization: String,
    pub common: String,
    pub days: Option<i64>,
    pub sig_algo: CertSigAlgo,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CertSigAlgo {
    EcDsa,
    ED25519,
}

impl CertInfo {
    pub fn new(
        domains: Vec<impl Into<String>>,
        ips: Vec<impl Into<String>>,
        country: &str,
        org: &str,
        cn: &str,
        days: Option<i64>,
        sig_algo: CertSigAlgo,
    ) -> Self {
        Self {
            domain_names: domains.into_iter().map(|d| d.into()).collect(),
            ip_address: ips
                .into_iter()
                .map(|ip| ip.into().parse().unwrap())
                .collect(),
            country: country.to_owned(),
            organization: org.to_owned(),
            common: cn.to_owned(),
            days,
            sig_algo,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CertType {
    Client,
    Server,
    CA,
}

impl CertInfo {
    fn build_cert_params(
        &self,
        keypair: Option<KeyPair>,
        cert_type: CertType,
    ) -> CertificateParams {
        let alg = self.sig_algo.into();

        let default_days = if cert_type == CertType::CA {
            CA_DEFAULT_DURATION
        } else {
            CERT_DEFAULT_DURATION
        };
        let days = self.days.unwrap_or(default_days);

        let not_before = OffsetDateTime::now_utc();

        let not_after = not_before + Duration::days(days);

        let mut subject_alt_names = vec![];
        for dns in self.domain_names.iter() {
            subject_alt_names.push(SanType::DnsName(dns.to_owned()));
        }
        for ip in self.ip_address.iter() {
            subject_alt_names.push(SanType::IpAddress(ip.to_owned()));
        }
        let mut distinguished_name = DistinguishedName::new();
        distinguished_name.push(DnType::CountryName, &self.country);
        distinguished_name.push(DnType::OrganizationName, &self.organization);
        distinguished_name.push(DnType::CommonName, &self.common);

        let mut params = CertificateParams::default();
        params.key_pair = keypair;
        params.alg = alg;
        params.not_before = not_before;
        params.not_after = not_after;
        params.subject_alt_names = subject_alt_names;
        params.distinguished_name = distinguished_name;

        // WTF: turned out we should not use this to satisfy openssl
        // params.use_authority_key_identifier_extension = true;
        params.key_identifier_method = KeyIdMethod::Sha512;
        match cert_type {
            CertType::CA => {
                params.extended_key_usages = vec![Any];
                params.is_ca = IsCa::Ca(BasicConstraints::Constrained(16));
                params.custom_extensions.push(CertInfo::key_usage(true));
            }
            CertType::Client => {
                params.extended_key_usages = vec![ClientAuth];
                params.custom_extensions.push(CertInfo::not_ca());
                params.custom_extensions.push(CertInfo::key_usage(false));
            }
            CertType::Server => {
                params.extended_key_usages = vec![ServerAuth];
                params.custom_extensions.push(CertInfo::not_ca());
                params.custom_extensions.push(CertInfo::key_usage(false));
            }
        }

        params
    }

    pub fn ca_cert(&self, keypair: Option<KeyPair>) -> Result<CA, CertifyError> {
        let mut params = self.build_cert_params(keypair, CertType::CA);
        params.extended_key_usages = vec![];
        CA::from_params(params)
    }

    pub fn client_cert(&self, keypair: Option<KeyPair>) -> Result<Cert, CertifyError> {
        let params = self.build_cert_params(keypair, CertType::Client);
        Cert::from_params(params)
    }

    pub fn server_cert(&self, keypair: Option<KeyPair>) -> Result<Cert, CertifyError> {
        let params = self.build_cert_params(keypair, CertType::Server);
        Cert::from_params(params)
    }

    fn key_usage(ca: bool) -> CustomExtension {
        let der = yasna::construct_der(|writer| {
            writer.write_bitvec(
                &KEY_USAGE
                    .iter()
                    .map(|u| *u == if ca { Usage::Ca } else { Usage::Cert })
                    .collect(),
            );
        });

        let mut key_usage = CustomExtension::from_oid_content(OID_KEY_USAGE, der);
        key_usage.set_criticality(true);
        key_usage
    }

    fn not_ca() -> CustomExtension {
        let der = yasna::construct_der(|writer| {
            writer.write_sequence(|writer| {
                writer.next().write_bool(false);
            });
        });

        CustomExtension::from_oid_content(OID_BASIC, der)
    }
}

impl From<CertSigAlgo> for &SignatureAlgorithm {
    fn from(algo: CertSigAlgo) -> Self {
        match algo {
            CertSigAlgo::EcDsa => &PKCS_ECDSA_P384_SHA384,
            CertSigAlgo::ED25519 => &PKCS_ED25519,
        }
    }
}
